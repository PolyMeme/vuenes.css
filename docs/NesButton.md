# nes-button 

Vue Component implementation of the Nes Button 

- **author** - Jesus Fernando Alvarez Franco 
- **license** - MIT 

## slots 

- `default` Use this slot for the content Inside the button 

## props 

- `primary` ***Boolean*** (*optional*) 

  Prop that define if the Button is type Primary. Will use the `is-primary` class 

- `success` ***Boolean*** (*optional*) 

  Prop that define if the Button is type Success. Will use the `is-success` class 

- `warning` ***Boolean*** (*optional*) 

  Prop that define if the Button is type Warning. Will use the `is-warning` class 

- `error` ***Boolean*** (*optional*) 

  Prop that define if the Button is type Error. Will use the `is-error` class 

- `disabled` ***Boolean*** (*optional*) 

  Prop that define if the Button is type Disabled. Will use the `is-disabled` class 

## computed properties 

- `buttonClasses` 

  Computed Property that uses the button classes depending of the props used. 

   **dependencies:** `primary`, `success`, `warning`, `error`, `disabled` 


## events 

- `click` 

  Fired when a the user clicks the button. 

## methods 

- `click()` 

  Method that emits the click event 

