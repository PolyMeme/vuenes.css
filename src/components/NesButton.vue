<template>
  <button
    class="nes-btn"
    :class="buttonClasses"
    type="button"
    @click="click"
  >
    <!-- Use this slot for the content Inside the button -->
    <slot />
  </button>
</template>

<script>
/**
 * Vue Component implementation of the Nes Button
 * @author Jesus Fernando Alvarez Franco
 * @license MIT
 */
export default {
  props: {
    /**
     * Prop that define if the Button is type Primary. Will use the `is-primary` class
     * @default false
     */
    primary: Boolean,
    /**
     * Prop that define if the Button is type Success. Will use the `is-success` class
     * @default false
     */
    success: Boolean,
    /**
     * Prop that define if the Button is type Warning. Will use the `is-warning` class
     * @default false
     */
    warning: Boolean,
    /**
     * Prop that define if the Button is type Error. Will use the `is-error` class
     * @default false
     */
    error: Boolean,
    /**
     * Prop that define if the Button is type Disabled. Will use the `is-disabled` class and will not trigger the click event
     * @default false
     */
    disabled: Boolean
  },
  computed: {
    /**
     * Computed Property that uses the button classes depending of the props used.
     * @returns {{'is-primary': BooleanConstructor | *, 'is-error': BooleanConstructor | *, 'is-success': BooleanConstructor | *, 'is-disabled': BooleanConstructor | *, 'is-warning': BooleanConstructor | *}}
     */
    buttonClasses () {
      return {
        'is-primary': this.primary,
        'is-success': this.success,
        'is-warning': this.warning,
        'is-error': this.error,
        'is-disabled': this.disabled
      }
    }
  },
  methods: {
    /**
     * Method that emits the click event
     */
    click () {
      /**
       * Fired when a the user clicks the button and the button is not disabled
       */
      if (!this.disabled) {
        this.$emit('click')
      }
    }
  }
}
</script>
