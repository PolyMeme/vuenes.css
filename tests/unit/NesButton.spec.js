import { mount } from '@vue/test-utils'
import NesButton from '../../src/components/NesButton.vue'

describe('NesButton.vue', () => {
  const createButtonWrapper = (propsData = {}, slot) => {
    return mount(NesButton, {
      propsData,
      slots: {
        default: slot || 'I am a button'
      }
    })
  }

  it('should render button with proper button tag', () => {
    const wrapper = createButtonWrapper()

    expect(wrapper.exists()).toBe(true)
    expect(wrapper.find('button').exists()).toBe(true)
    expect(wrapper.text()).toBe('I am a button')
  })

  it('should render button with proper button tag', () => {
    const wrapper = createButtonWrapper()
    const buttonClasses = wrapper.classes()

    expect(buttonClasses.length).toBe(1)
    expect(buttonClasses).toContainEqual('btn')
  })

  it('should emit click event', () => {
    const wrapper = createButtonWrapper()

    // Trigger Click and Emit Three Times
    wrapper.find('button.btn').trigger('click')
    wrapper.find('button.btn').trigger('click')
    wrapper.find('button.btn').trigger('click')

    expect(wrapper.emitted('click')).toBeTruthy()
    expect(wrapper.emitted('click').length).toBe(3)
  })

  it('should add \'is-primary\' class if primary prop is added', () => {
    const wrapper = createButtonWrapper({
      primary: true
    })
    const buttonClasses = wrapper.classes()

    expect(buttonClasses.length).toBe(2)
    expect(buttonClasses).toContainEqual('is-primary')
  })

  it('should add \'is-success\' class if success prop is added', () => {
    const wrapper = createButtonWrapper({
      success: true
    })
    const buttonClasses = wrapper.classes()

    expect(buttonClasses.length).toBe(2)
    expect(buttonClasses).toContainEqual('is-success')
  })

  it('should add \'is-warning\' class if warning prop is added', () => {
    const wrapper = createButtonWrapper({
      warning: true
    })
    const buttonClasses = wrapper.classes()

    expect(buttonClasses.length).toBe(2)
    expect(buttonClasses).toContainEqual('is-warning')
  })

  it('should add \'is-error\' class if error prop is added', () => {
    const wrapper = createButtonWrapper({
      error: true
    })
    const buttonClasses = wrapper.classes()

    expect(buttonClasses.length).toBe(2)
    expect(buttonClasses).toContainEqual('is-error')
  })

  it('should print html inside slot', () => {
    const wrapper = createButtonWrapper({}, '<p>Hello there!</p>')
    const slot = wrapper.find('button>p')

    expect(wrapper.exists()).toBe(true)
    expect(slot.exists()).toBe(true)
    expect(slot.html()).toBe('<p>Hello there!</p>')
  })

  it('should match snapshot', () => {
    const wrapper = createButtonWrapper({
      success: true
    }, '<h3>I am a snapshot!</h3>')

    expect(wrapper.html()).toMatchSnapshot()
  })
})
