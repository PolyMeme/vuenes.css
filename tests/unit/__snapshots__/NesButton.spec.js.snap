// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NesButton.vue should match snapshot 1`] = `
<button type="button" class="nes-btn is-success">
  <h3>I am a snapshot!</h3>
</button>
`;
